tests/test_smoke.py

```diff
@@ -2606,6 +2612,280 @@ def test_gcp_zero_quota_failover():
+                                    serve.LB_CONTROLLER_SYNC_INTERVAL + 5)
+
+    test = Test(
+        f'test-skyserve-replica-failure',
+        [
+            f'sky serve up -n {name} -y tests/skyserve/replica_failure/service.yaml',
+            _SERVE_WAIT_UNTIL_READY.format(name=name, replica_num=3),
+            f'{_get_serve_endpoint(name)}; {_get_replica_ip(name, 1)}; '
+            f'{_get_replica_ip(name, 2)}; {_get_replica_ip(name, 3)}; '
+            'python tests/skyserve/replica_failure/test_round_robin.py '
+            '--endpoint $endpoint --replica-num 3 --replica-ips $ip1 $ip2 $ip3',
+            terminate_replica(1),
+            f'sleep {time_to_wait_after_terminate}',
+            f'sky serve status {name} | grep 2/3',
+            f'{_get_replica_line(name, 1)} | grep NOT_READY',
+            f'{_get_serve_endpoint(name)}; {_get_replica_ip(name, 2)}; '
+            f'{_get_replica_ip(name, 3)}; '
+            'python tests/skyserve/replica_failure/test_round_robin.py '
+            '--endpoint $endpoint --replica-num 2 --replica-ips $ip2 $ip3',
+            terminate_replica(2),
+            f'sleep {time_to_wait_after_terminate}',
+            f'sky serve status {name} | grep 1/3',
+            f'{_get_replica_line(name, 2)} | grep NOT_READY',
+            f'{_get_serve_endpoint(name)}; {_get_replica_ip(name, 3)}; '
+            'python tests/skyserve/replica_failure/test_round_robin.py '
+            '--endpoint $endpoint --replica-num 1 --replica-ips $ip3',
+        ],
+        f'sky serve down -y {name}',
+        timeout=20 * 60,
+    )
+    run_one_test(test)
+
+
+@pytest.mark.gcp
+@pytest.mark.sky_serve
+def test_skyserve_auto_restart():
+    """Test skyserve with auto restart"""
+    name = _get_service_name()
+    zone = 'us-central1-a'
+
+    # Reference: test_spot_recovery_gcp
+    def terminate_replica(replica_id: int) -> str:
+        cluster_name = serve.generate_replica_cluster_name(name, replica_id)
+        query_cmd = (f'gcloud compute instances list --filter='
+                     f'"(labels.ray-cluster-name:{cluster_name})" '
+                     f'--zones={zone} --format="value(name)"')
+        return (f'gcloud compute instances delete --zone={zone}'
+                f' --quiet $({query_cmd})')
+
+    test = Test(
+        f'test-skyserve-auto-restart',
+        [
+            # TODO(tian): we can dynamically generate YAML from template to
+            # avoid maintaining too many YAML files
+            f'sky serve up -n {name} -y tests/skyserve/auto_restart.yaml',
+            _SERVE_WAIT_UNTIL_READY.format(name=name, replica_num=1),
+            f'{_get_serve_endpoint(name)}; curl -L http://$endpoint | grep "Hi, SkyPilot here"',
+            # sleep for 20 seconds (initial delay) to make sure it will
+            # be restarted
+            f'sleep 20',
+            terminate_replica(1),
+            'sleep 180',  # Wait for consecutive failure timeout passed.
```

Comment:
Because in the normal case, it won't check the cluster status on the cloud (since manual shutdown is not a common behavior and such queroes takes a lot of time). Instead, we think continuous 3 min probe failure is not a temporary problem but indeed a failure.

Reply:
Worth a comment